proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -70,14 +70,65 @@ primrec sc_ctrl_inv_rel ::
+   decodeSchedContext_Bind sc excaps
+   \<lbrace>valid_sc_inv'\<rbrace>, -"
+  apply (clarsimp simp: decodeSchedContext_Bind_def)
+  apply (wpsimp wp: gts_wp' threadGet_wp getNotification_wp
+              simp: scReleased_def scActive_def isBlocked_def refillReady_def)
+  apply (clarsimp simp: valid_cap'_def)
+  apply (drule_tac x="hd excaps" in bspec, fastforce dest: hd_in_set)+
+  apply (fastforce simp: pred_tcb_at'_def obj_at'_def)
+  done
+
+lemma decodeSchedContext_UnbindObject_wf:
+  "\<lbrace>valid_cap' (SchedContextCap sc n) and ex_nonz_cap_to' sc
+    and (\<lambda>s. \<forall>cap\<in>set excaps. \<forall>r\<in>zobj_refs' cap. ex_nonz_cap_to' r s)
+    and (\<lambda>s. \<forall>x\<in>set excaps. valid_cap' x s)\<rbrace>
+   decodeSchedContext_UnbindObject sc excaps
+   \<lbrace>valid_sc_inv'\<rbrace>, -"
+  apply (clarsimp simp: decodeSchedContext_UnbindObject_def)
+  apply (wpsimp wp: gts_wp' threadGet_wp getNotification_wp
+              simp: scReleased_def scActive_def isBlocked_def refillReady_def)
+  apply (clarsimp simp: valid_cap'_def)
+  apply (drule_tac x="hd excaps" in bspec, fastforce dest: hd_in_set)+
+  apply (fastforce simp: pred_tcb_at'_def obj_at'_def)
+  done
+
+lemma decodeSchedContext_YieldTo_wf:
+  "\<lbrace>valid_cap' (SchedContextCap sc n) and ex_nonz_cap_to' sc\<rbrace>
+   decodeSchedContext_YieldTo sc args
+   \<lbrace>valid_sc_inv'\<rbrace>, -"
+  apply (clarsimp simp: decodeSchedContext_YieldTo_def)
+  apply (wpsimp wp: gts_wp' threadGet_wp getNotification_wp getTCB_wp
+              simp: scReleased_def scActive_def isBlocked_def refillReady_def)
+  apply (clarsimp simp: valid_cap'_def)
+  apply (clarsimp simp: pred_tcb_at'_def obj_at'_def projectKOs)
+  done
+
+lemma valid_SchedContextCap_sc_at':
+  "\<lbrakk>valid_cap' (capability.SchedContextCap sc n) s\<rbrakk> \<Longrightarrow> sc_at' sc s"
+  apply (clarsimp simp: valid_cap'_def obj_at'_real_def)
+  apply (rule ko_wp_at'_weakenE)
+   apply (fastforce simp: objBits_simps
+                   split: kernel_object.splits)+
+  done
+
 lemma decodeSchedContextInvocation_wf:
-  "\<lbrace> valid_cap' (SchedContextCap sc n) and invs' and sch_act_simple and ex_nonz_cap_to' sc and
-     (\<lambda>s. \<forall>cap\<in>set excaps. \<forall>r\<in>zobj_refs' cap. ex_nonz_cap_to' r s) and
-     (\<lambda>s. \<forall>x\<in>set excaps. valid_cap' x s) \<rbrace>
+  "\<lbrace>valid_cap' (SchedContextCap sc n) and ex_nonz_cap_to' sc
+    and (\<lambda>s. \<forall>cap\<in>set excaps. \<forall>r\<in>zobj_refs' cap. ex_nonz_cap_to' r s)
+    and (\<lambda>s. \<forall>x\<in>set excaps. valid_cap' x s)\<rbrace>
    decodeSchedContextInvocation label sc excaps args
    \<lbrace>valid_sc_inv'\<rbrace>, -"
-  sorry
+  apply (simp add: decodeSchedContextInvocation_def)
+  apply (wpsimp wp: decodeSchedContext_Bind_wf[where n=n]
+                    decodeSchedContext_UnbindObject_wf[where n=n]
+                    decodeSchedContext_YieldTo_wf[where n=n])
```

Comment:
It sometimes ends up causing more problems than it helps, but you can avoid having to specify `n` here by changing these lemmas to be `∃n. valid_cap' (SchedContextCap sc n) ...`